src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java

```diff
@@ -320,7 +463,233 @@ private void updateWorkOrders()
+
+        if (pos != null)
+        {
+            MineColonies.getNetwork().sendToServer(new ChangeFreeToInteractBlockMessage(townHall.getColony(), pos, ChangeFreeToInteractBlockMessage.MessageType.ADD_BLOCK));
+            townHall.getColony().addFreePosition(pos);
+        }
+
+        fillFreeBlockList();
+        input.setText("");
+    }
+
+    /**
+     * Called when the permission button has been triggered.
+     *
+     * @param button the triggered button.
+     */
+    private void trigger(@NotNull final Button button)
+    {
+        @NotNull final Pane pane = button.getParent().getChildren().get(2);
+        int index = 0;
+        if (pane instanceof Label)
+        {
+            index = Integer.valueOf(((Label) pane).getLabelText());
+        }
+        final boolean trigger = LanguageHandler.format(ON).equals(button.getLabel());
+        final Permissions.Action action = Permissions.Action.values()[index];
+        final Permissions.Rank rank = Permissions.Rank.valueOf(actionsList.getParent().getID().toUpperCase());
+
+        MineColonies.getNetwork().sendToServer(new PermissionsMessage.Permission(townHall.getColony(), PermissionsMessage.MessageType.TOGGLE_PERMISSION, rank, action));
+        townHall.getColony().getPermissions().togglePermission(rank, action);
+
+        if (trigger)
+        {
+            button.setLabel(LanguageHandler.format(OFF));
+        }
+        else
+        {
+            button.setLabel(LanguageHandler.format(ON));
+        }
+    }
+
+    /**
+     * Switch between previous and next page.
+     */
+    private void switchPage(@NotNull final Button button)
+    {
+        if (button.getID().equals(BUTTON_PREV_PAGE_PERM))
+        {
+            findPaneOfTypeByID(VIEW_PERM_PAGES, SwitchView.class).previousView();
+
+            findPaneOfTypeByID(BUTTON_PREV_PAGE_PERM, Button.class).setEnabled(false);
+            findPaneOfTypeByID(BUTTON_NEXT_PAGE_PERM, Button.class).setEnabled(true);
+        }
+        else
+        {
+            findPaneOfTypeByID(VIEW_PERM_PAGES, SwitchView.class).nextView();
+
+            findPaneOfTypeByID(BUTTON_PREV_PAGE_PERM, Button.class).setEnabled(true);
+            findPaneOfTypeByID(BUTTON_NEXT_PAGE_PERM, Button.class).setEnabled(false);
+        }
+
+        if (findPaneOfTypeByID(VIEW_PERM_PAGES, SwitchView.class).getCurrentView().getID().equals(PERMISSION_VIEW))
+        {
+            findPaneOfTypeByID(BUTTON_PREV_PAGE_PERM, Button.class).setEnabled(true);
+            findPaneOfTypeByID(BUTTON_NEXT_PAGE_PERM, Button.class).setEnabled(true);
+
+            fillPermissionList(VIEW_OFFICER);
+        }
+    }
+
+    /**
+     * Fills the permission list in the GUI.
+     */
+    private void fillPermissionList(@NotNull final String category)
+    {
+        actionsList = findPaneOfTypeByID(LIST_ACTIONS + category, ScrollingList.class);
+        actionsList.setDataProvider(new ScrollingList.DataProvider()
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Reduce this anonymous class number of lines from 24 to at most 20, or make it a named class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)